### PR TITLE
Initial impl of RFC-277 - native exe throws on non-zero exit code

### PR DESCRIPTION
**Pull request by rkeithhill**
# PR Summary

Implements a terminating error when a native executable returns a non-zero exit code similar to `set -e` on Linux and macOS.

## PR Context
Implement [RFC 277](https://github.com/PowerShell/PowerShell-RFC/pull/277)

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [ ] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [x] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/staging/reference/6/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): 
- **User-facing changes**
    - [ ] Not Applicable
    - **OR**
    - [x] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: 
- **Testing - New and feature**
    - [ ] N/A or can only be tested interactively
    - **OR**
    - [ ] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [x] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and opened an issue in the relevant tool repository. This may include:
        - [ ] Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode
        (which runs in a different PS Host).
            - [ ] Issue filed: 
        - [ ] Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
            - [ ] Issue filed: 
        - [x] Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
            - [ ] Issue filed: 
        - [ ] Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).
            - [ ] Issue filed: 

This PR does impact the [UseDeclaredVarsMoreThanAssignments PSSA rule](https://github.com/PowerShell/PSScriptAnalyzer/blob/master/RuleDocumentation/UseDeclaredVarsMoreThanAssignments.md).  `PSNativeCommandUseErrorActionPreference` needs to be excluded from this rule.

### CODE REVIEW

File: src/System.Management.Automation/engine/NativeCommandProcessor.cs [hunk 1]
```diff
@@ -726,7 +726,26 @@ internal override void Complete()
 
                     this.Command.Context.SetVariable(SpecialVariables.LastExitCodeVarPath, _nativeProcess.ExitCode);
                     if (_nativeProcess.ExitCode != 0)
+                    {
                         this.commandRuntime.PipelineProcessor.ExecutionFailed = true;
```
Comment (rkeithhill) [inline src/System.Management.Automation/engine/NativeCommandProcessor.cs:730]: If we throw a terminating error, do we need to execute this line?  If not, I could move this line after the test if we should throw.

File: src/System.Management.Automation/engine/SpecialVariables.cs [hunk 2]
```diff
@@ -406,5 +414,6 @@ internal enum PreferenceVariable
         Warning = 13,
         Information = 14,
         Confirm = 15,
+        NativeCommandThrow = 16,
```
Comment (rkeithhill) [inline src/System.Management.Automation/engine/SpecialVariables.cs:417]: I'm not sure if this preference needs to be defined here.

**Comment by rkeithhill** [comment]
There are internal APIs to get/set pref vars in `ExecutionContext.cs`, `MshCommandRuntime.cs` and `PowerShell.cs` (public api). Not sure if I need to define this preference variable in those files.

File: src/System.Management.Automation/engine/NativeCommandProcessor.cs (hunk 1 above)
Comment (iSazonov) [reply src/System.Management.Automation/engine/NativeCommandProcessor.cs:730]: This depends on NativeCommandThrowPreference. If PSNativeCommandUseErrorActionPreference  we keep old behavior.

**Comment by rjmholt** [comment]
Assigning myself since I've been tasked with taking this work on, but since you've already started @rkeithhill maybe I can support you here instead

**Comment by rkeithhill** [comment]
@rjmholt That would be great! I'd love to have your input on this.  I need to update this PR as I started it off of my memory of the RFC and now I see that (as @iSazonov  points out), I need to be checking the value of `$ErrorActionPreference`.  I also see in the PR that there are specifics regarding the generated `ErrorRecord` that I need to snap to.  I'll push an update to this PR tonight to cover those issues.  I'll try to add the implementation for `NativeCommandException` but may have some questions about where to put that source.

File: src/System.Management.Automation/engine/NativeCommandProcessor.cs [hunk 3]
```diff
@@ -726,7 +726,37 @@ internal override void Complete()
 
                     this.Command.Context.SetVariable(SpecialVariables.LastExitCodeVarPath, _nativeProcess.ExitCode);
                     if (_nativeProcess.ExitCode != 0)
```
Comment (rkeithhill) [inline src/System.Management.Automation/engine/NativeCommandProcessor.cs:728]: @rjmholt The code after this code is basically just for handling exceptions (and finally cleanup).  I think this code would read easier (less indentation) if I switched up to early return instead of nesting. Thoughts?

File: src/System.Management.Automation/engine/SpecialVariables.cs [hunk 4]
```diff
@@ -255,6 +255,11 @@ internal static class SpecialVariables
 
         internal static readonly VariablePath InformationPreferenceVarPath = new VariablePath(InformationPreference);
 
+        internal const string NativeCommandUseErrorActionPreference = "PSNativeCommandUseErrorActionPreference";
```
Comment (rkeithhill) [inline src/System.Management.Automation/engine/SpecialVariables.cs:258]: This is a lengthy variable name but I suppose it's very descriptive / intentional.

File: src/System.Management.Automation/resources/RunspaceInit.resx [hunk 5]
```diff
@@ -186,6 +186,9 @@
   <data name="NestedPromptLevelDescription" xml:space="preserve">
     <value>Dictates what type of prompt should be displayed for the current nesting level</value>
   </data>
+  <data name="NativeCommandUseErrorActionPreferenceDescription" xml:space="preserve">
+    <value>If true, $ErrorActionPreference applies to native executable non-zero exit codes</value>
```
Comment (rkeithhill) [inline src/System.Management.Automation/resources/RunspaceInit.resx:190]: Could use some help with this wording.

File: src/System.Management.Automation/engine/NativeCommandProcessor.cs (hunk 1 above)
Comment (rkeithhill) [reply src/System.Management.Automation/engine/NativeCommandProcessor.cs:730]: I think my question does not apply anymore since the change I made.  This code will now only `Write` an ErrorRecord and not call ThrowTerminatingError().

File: src/System.Management.Automation/engine/NativeCommandProcessor.cs [hunk 6]
```diff
@@ -726,7 +726,36 @@ internal override void Complete()
...
+                    {
                         this.commandRuntime.PipelineProcessor.ExecutionFailed = true;
+
+                        bool nativeCommandUseErrorActionPref = 
+                            this.Command.Context.GetBooleanPreference(
+                                SpecialVariables.NativeCommandUseErrorActionPreferenceVarPath, 
+                                false, 
+                                out _);
```
Comment (iSazonov) [inline src/System.Management.Automation/engine/NativeCommandProcessor.cs:736]:
```suggestion
                        bool nativeCommandUseErrorActionPreference = 
                            this.Command.Context.GetBooleanPreference(
                                SpecialVariables.NativeCommandUseErrorActionPreferenceVarPath, 
                                InitialSessionState.DefaultNativeCommandUseErrorActionPreference, 
                                out _);
```

File: src/System.Management.Automation/engine/NativeCommandProcessor.cs [hunk 7]
```diff
@@ -726,7 +726,36 @@ internal override void Complete()
...
+
+                        // If $PSNativeCommandUseErrorActionPreference is $true, then evaluate $ErrorActionPreference
```
Comment (iSazonov) [inline src/System.Management.Automation/engine/NativeCommandProcessor.cs:738]: Please remove obvious comment.

File: src/System.Management.Automation/engine/NativeCommandProcessor.cs [hunk 8]
```diff
@@ -726,7 +726,36 @@ internal override void Complete()
...
+
+                            if (errorActionPref != ActionPreference.Ignore)
+                            {
+                                var errorRecord = 
+                                    new ErrorRecord(
+                                        exception: new Exception($"Program {this.NativeCommandName} ended with non-zero exit code {_nativeProcess.ExitCode}"), 
+                                        errorId: "NativeCommandThrow", 
+                                        errorCategory: ErrorCategory.NotSpecified,
+                                        targetObject: this.NativeCommandName);
+                                this.commandRuntime._WriteErrorSkipAllowCheck(errorRecord, isNativeError: true);
```
Comment (iSazonov) [inline src/System.Management.Automation/engine/NativeCommandProcessor.cs:755]: Since we have isNativeError flag why not move the new code to _WriteErrorSkipAllowCheck()?

Comment (rkeithhill) [reply src/System.Management.Automation/engine/NativeCommandProcessor.cs:755]: I'll look into that.

File: src/System.Management.Automation/engine/SpecialVariables.cs (hunk 4 above)
Comment (rjmholt) [reply src/System.Management.Automation/engine/SpecialVariables.cs:258]: Yeah I don't love the variable name prescribed by the RFC, but we're bound by our process there I guess.

In the [small amount of work I'd done toward this](https://github.com/rjmholt/PowerShell/commit/bab0480be7a0403c9fdce0a5307dd2cca36af854), I used `nameof` to deduplicate the variable name.

File: src/System.Management.Automation/engine/NativeCommandProcessor.cs (hunk 3 above)
Comment (rjmholt) [reply src/System.Management.Automation/engine/NativeCommandProcessor.cs:729]: Absolutely agreed. I want to hold fewer states in my head as I read down.

File: src/System.Management.Automation/engine/NativeCommandProcessor.cs (hunk 8 above)
Comment (rjmholt) [reply src/System.Management.Automation/engine/NativeCommandProcessor.cs:755]: Yeah this was kind of a complicated question to answer, but I think it's worth us taking a look

File: src/System.Management.Automation/resources/RunspaceInit.resx (hunk 5 above)
Comment (rjmholt) [reply src/System.Management.Automation/resources/RunspaceInit.resx:190]: I think what you've got is pretty good -- hard to word. One alternative:

> If true, $ErrorActionPreference and -ErrorAction parameters will apply to native executables, so that non-zero exit codes will generate cmdlet-style errors governed by error action settings.

File: src/System.Management.Automation/engine/NativeCommandProcessor.cs (hunk 3 above)
Comment (rkeithhill) [reply src/System.Management.Automation/engine/NativeCommandProcessor.cs]: Done

**Comment by rkeithhill** [comment]
I implemented `NativeCommandException`. The result looks like this:

image

File: src/System.Management.Automation/resources/CommandBaseStrings.resx [hunk 9]
```diff
@@ -156,6 +156,9 @@
   <data name="PauseHelpMessage" xml:space="preserve">
     <value>Pause the current pipeline and return to the command prompt. Type "{0}" to resume the pipeline.</value>
   </data>
+  <data name="ProgramFailedToComplete" xml:space="preserve">
```
Comment (rkeithhill) [inline src/System.Management.Automation/resources/CommandBaseStrings.resx:159]: There is a ProgramFailedToExecute resource string in `ParserStrings.resx` but that doesn't seem to be quite right for this runtime error.  Is `CommandBaseStrings.resx` the correct file for this resource sting?

File: src/System.Management.Automation/engine/SpecialVariables.cs (hunk 4 above)
Comment (rkeithhill) [reply src/System.Management.Automation/engine/SpecialVariables.cs]: Now using `nameof()`

File: src/System.Management.Automation/resources/RunspaceInit.resx (hunk 5 above)
Comment (rkeithhill) [reply src/System.Management.Automation/resources/RunspaceInit.resx:190]: @rjmholt Can you give an example of where the `-ErrorAction` parameter would come into play? I've tried `iex "whoami -xyzzy" -ea Stop` and `icm { whoami -xyzzy } -ea Stop`.  If I check the value of `$ErrorActionPreference` with the iex string or the icm scriptbock, the `-ErrorAction` parameter does not impact the value or pref variable.
```
PS> icm {
     "`$ErrorActionPreference is $ErrorActionPreference"
     $PSNativeCommandUseErrorActionPreference=$true
     whoami -xyzzy
} -ErrorAction Stop
$ErrorActionPreference is Continue
ERROR: Invalid argument/option - '-xyzzy'.
Type "WHOAMI /?" for usage.

NativeCommandException: Program "whoami.exe" ended with non-zero exit code 1.
```

File: src/System.Management.Automation/resources/CommandBaseStrings.resx (hunk 9 above)
Comment (iSazonov) [reply src/System.Management.Automation/resources/CommandBaseStrings.resx:159]: It seems it is not used at all.

Comment (rkeithhill) [reply src/System.Management.Automation/resources/CommandBaseStrings.resx:159]: It is used in the same NativeCommandProcessor.cs file during the creation of another exception `ApplicationFailedException` that I thiink covers the general case of native command execution failure.  This does make me wonder if someone looking at the code in the future is going to wonder why there's both an `ApplicationFailedException` and a `NativeCommandException`.

File: test/powershell/engine/Basic/NativeCommandErrorHandling.Tests.ps1 [hunk 10]
```diff
@@ -0,0 +1,143 @@
+# Copyright (c) Microsoft Corporation.
+# Licensed under the MIT License.
+#
+# Functional tests to verify that errors (non-terminating and terminating) appropriately
+# when $PSNativeCommandUseErrorActionPreference is $true
+
+Describe 'Native command error handling tests' -Tags 'CI' {
+    Context 'PSNativeCommandUseErrorActionPreference is $true' {
```
Comment (iSazonov) [inline test/powershell/engine/Basic/NativeCommandErrorHandling.Tests.ps1:8]: I'd combine tests for true and false (and maybe for $ErrorActionPreference values) and use TestCases.

File: src/System.Management.Automation/resources/CommandBaseStrings.resx (hunk 9 above)
Comment (iSazonov) [reply src/System.Management.Automation/resources/CommandBaseStrings.resx:159]: There is still used ParserStrings.CantActivateDocumentInPowerShellCore with InterpreterError.NewInterpreterException.

File: src/System.Management.Automation/resources/RunspaceInit.resx (hunk 5 above)
Comment (rjmholt) [reply src/System.Management.Automation/resources/RunspaceInit.resx:190]: No you're right, there's no `-ErrorAction` parameter here. Just the preference variable

**Comment by rkeithhill** [comment]
@rjmholt The current build checks seem to be stuck. Is there a way I restart them short of pushing a bogus commit?

**Comment by rjmholt** [comment]
Test failures are very weird...

/azp run

**Comment by rkeithhill** [comment]
@rjmholt There is a test failure in output rendering tests which I don't believe this PR should impact:
```
2021-08-05T17:44:15.0467419Z     Context NormalView tests
2021-08-05T17:44:15.0498743Z +
2021-08-05T17:44:15.0512900Z 
2021-08-05T17:44:15.0527114Z     Context DetailedView tests
2021-08-05T17:44:15.0555317Z +
2021-08-05T17:44:15.0570773Z 
2021-08-05T17:44:15.0603255Z 
2021-08-05T17:44:15.0611669Z   Describing OutputRendering tests
2021-08-05T17:44:15.0627692Z     [-] Error occurred in Describe block
2021-08-05T17:44:15.0649066Z       PSArgumentException: The key 'It:Skip' has already been added to the dictionary.
2021-08-05T17:44:15.0663568Z       MethodInvocationException: Exception calling "Add" with "2" argument(s): "The key 'It:Skip' has already been added to the dictionary."
2021-08-05T17:44:15.0704752Z       at <ScriptBlock>, D:\a\1\s\test\powershell\engine\Formatting\OutputRendering.Tests.ps1: line 6
2021-08-05T17:44:15.0717916Z       at Invoke-Blocks, D:\a\1\s\src\powershell-win-core\bin\Release\net6.0\win7-x64\publish\Modules\Pester\4.10.1\Functions\SetupTeardown.ps1: line 135
2021-08-05T17:44:15.0731656Z       at Invoke-TestGroupSetupBlocks, D:\a\1\s\src\powershell-win-core\bin\Release\net6.0\win7-x64\publish\Modules\Pester\4.10.1\Functions\SetupTeardown.ps1: line 121
2021-08-05T17:44:15.0746462Z       at DescribeImpl, D:\a\1\s\src\powershell-win-core\bin\Release\net6.0\win7-x64\publish\Modules\Pester\4.10.1\Functions\Describe.ps1: line 209
```
It's balking on this last line here:
```
Describe 'OutputRendering tests' {
    BeforeAll {
        $PSDefaultParameterValues.Add('It:Skip', (-not $EnabledExperimentalFeatures.Contains('PSAnsiRendering')))
```
Anything wrong with that? It's a bit different than how you skipped the tests for this PR when the exp feature is not enabled.

File: src/System.Management.Automation/engine/SpecialVariables.cs [hunk 11]
```diff
@@ -321,25 +326,29 @@ internal static class SpecialVariables
...
+        internal static readonly Type[] PreferenceVariableTypes =
+        {
+            /* DebugPreference */                         typeof(ActionPreference),
+            /* VerbosePreference */                       typeof(ActionPreference),
+            /* ErrorPreference */                         typeof(ActionPreference),
+            /* WhatIfPreference */                        typeof(SwitchParameter),
+            /* WarningPreference */                       typeof(ActionPreference),
+            /* InformationPreference */                   typeof(ActionPreference),
+            /* ConfirmPreference */                       typeof(ConfirmImpact),
+            /* PSNativeCommandUseErrorActionPreference */ typeof(SwitchParameter),
```
Comment (rjmholt) [inline src/System.Management.Automation/engine/SpecialVariables.cs:350]: This should be the same type as the builtin variable listing in InitialSessionState.cs

**Comment by rkeithhill** [comment]
@rjmholt Woohoo!  All checks pass!

File: src/System.Management.Automation/engine/SpecialVariables.cs (hunk 11 above)
Comment (rjmholt) [reply src/System.Management.Automation/engine/SpecialVariables.cs:350]: @rkeithhill

Comment (rjmholt) [reply src/System.Management.Automation/engine/SpecialVariables.cs:350]:
```suggestion
            /* PSNativeCommandUseErrorActionPreference */ typeof(bool),
```

File: src/System.Management.Automation/resources/CommandBaseStrings.resx (hunk 9 above)
Comment (rjmholt) [reply src/System.Management.Automation/resources/CommandBaseStrings.resx:159]: > This does make me wonder if someone looking at the code in the future is going to wonder why there's both an ApplicationFailedException and a NativeCommandException.

Worth adding a comment in the elements added in this PR to note that despite those things pre-existing, we added new ones to be specifically for the native error action preference feature to ensure it didn't accidentally clobber any existing concepts.

File: src/System.Management.Automation/resources/RunspaceInit.resx (hunk 5 above)
Comment (rjmholt) [reply src/System.Management.Automation/resources/RunspaceInit.resx:190]: Edited my suggestion above to correct it

File: test/powershell/engine/Basic/NativeCommandErrorHandling.Tests.ps1 (hunk 10 above)
Comment (rjmholt) [reply test/powershell/engine/Basic/NativeCommandErrorHandling.Tests.ps1:34]: The only things I would add here are:

- Testing on the properties of the exception that we're supposed to implement, if they're available beyond the error message
- A test with a try/catch
- A test with a trap

**Comment by theJasonHelmick** [comment]
@rkeithhill - Thank you for all the hard work. I just spoke with @rjmholt and he thinking he might open a separate pr to add a couple of tests.

File: src/System.Management.Automation/resources/CommandBaseStrings.resx (hunk 9 above)
Comment (rjmholt) [reply src/System.Management.Automation/resources/CommandBaseStrings.resx:159]: Adding this in my PR

File: test/powershell/engine/Basic/NativeCommandErrorHandling.Tests.ps1 (hunk 10 above)
Comment (rjmholt) [reply test/powershell/engine/Basic/NativeCommandErrorHandling.Tests.ps1:34]: Adding these in my PR

File: src/System.Management.Automation/engine/SpecialVariables.cs (hunk 11 above)
Comment (rjmholt) [reply src/System.Management.Automation/engine/SpecialVariables.cs:350]: Adding this in my PR

**Comment by rkeithhill** [comment]
Sounds good to me.  I wouldn't be able to do anything until later this evening and even then I don't have a ton of time before I'm out of vacation next week.

**Comment by rjmholt** [comment]
@rkeithhill see https://github.com/rkeithhill/PowerShell/pull/21

**Comment by rkeithhill** [comment]
Merged.

**Comment by rkeithhill** [comment]
Markdown tests failed. Le sigh. 

image

This is a downside of relying on websites during the build.

File: src/System.Management.Automation/engine/MshCommandRuntime.cs [hunk 12]
```diff
@@ -2819,7 +2819,7 @@ private void DoWriteError(object obj)
         /// but the command failure will ultimately be
         /// <see cref="System.Management.Automation.ActionPreferenceStopException"/>,
         /// </remarks>
-        internal void _WriteErrorSkipAllowCheck(ErrorRecord errorRecord, ActionPreference? actionPreference = null, bool isNativeError = false)
+        internal void _WriteErrorSkipAllowCheck(ErrorRecord errorRecord, ActionPreference? actionPreference = null, bool isNativeStdErrCall = false)
```
Comment (iSazonov) [inline src/System.Management.Automation/engine/MshCommandRuntime.cs:2822]: The change is not related to the PR. Please revert.

File: src/System.Management.Automation/engine/NativeCommandProcessor.cs [hunk 13]
```diff
@@ -762,8 +877,51 @@ internal override void Complete()
...
-                        this.commandRuntime.PipelineProcessor.ExecutionFailed = true;
+                    if (_nativeProcess.ExitCode == 0)
+                    {
+                        return;
+                    }
+
+                    this.commandRuntime.PipelineProcessor.ExecutionFailed = true;
+
+                    bool nativeCommandUseErrorActionPreference = ExperimentalFeature.IsEnabled(ExperimentalFeature.PSNativeCommandErrorActionPreferenceFeatureName)
+                        && this.Command.Context.GetBooleanPreference(
```
Comment (iSazonov) [inline src/System.Management.Automation/engine/NativeCommandProcessor.cs:888]: Why do we need the "this" prefix? Can we cleanup all in the PR (new and changed code only)?

File: src/System.Management.Automation/engine/NativeCommandProcessor.cs [hunk 14]
```diff
@@ -762,8 +877,51 @@ internal override void Complete()
...
+                            SpecialVariables.PSNativeCommandUseErrorActionPreferenceVarPath,
+                            InitialSessionState.DefaultPSNativeCommandUseErrorActionPreference,
+                            out _);
+
+                    if (!nativeCommandUseErrorActionPreference)
```
Comment (iSazonov) [inline src/System.Management.Automation/engine/NativeCommandProcessor.cs:893]: I'd prefer roll out the variable here and remove the help variable. This will simplify switching from experimental (less code changes).

File: src/System.Management.Automation/engine/NativeCommandProcessor.cs [hunk 15]
```diff
@@ -762,8 +877,51 @@ internal override void Complete()
...
+
+                    ActionPreference errorActionPref =
+                        this.Command.Context.GetEnumPreference(
+                            SpecialVariables.ErrorActionPreferenceVarPath,
+                            ActionPreference.SilentlyContinue,
+                            out _);
+
+                    if (errorActionPref != ActionPreference.Ignore)
+                    {
+                        const string errorId = "ProgramFailedToComplete";
```
Comment (iSazonov) [inline src/System.Management.Automation/engine/NativeCommandProcessor.cs:906]: Code convention for consts.
```suggestion
                        const string ErrorId = "ProgramFailedToComplete";
```

File: src/System.Management.Automation/engine/SpecialVariables.cs [hunk 16]
```diff
@@ -255,6 +255,11 @@ internal static class SpecialVariables
 
         internal static readonly VariablePath InformationPreferenceVarPath = new VariablePath(InformationPreference);
 
+        internal const string PSNativeCommandUseErrorActionPreference = nameof(PSNativeCommandUseErrorActionPreference);
```
Comment (iSazonov) [inline src/System.Management.Automation/engine/SpecialVariables.cs:258]: This can confuse a bit
```suggestion
        internal const string PSNativeCommandUseErrorActionPreferenceName = nameof(PSNativeCommandUseErrorActionPreference);
```

File: src/System.Management.Automation/engine/SpecialVariables.cs [hunk 17]
```diff
@@ -321,25 +326,29 @@ internal static class SpecialVariables
...
+        internal static readonly Type[] PreferenceVariableTypes =
+        {
+            /* DebugPreference */                         typeof(ActionPreference),
+            /* VerbosePreference */                       typeof(ActionPreference),
+            /* ErrorPreference */                         typeof(ActionPreference),
+            /* WhatIfPreference */                        typeof(SwitchParameter),
+            /* WarningPreference */                       typeof(ActionPreference),
+            /* InformationPreference */                   typeof(ActionPreference),
+            /* ConfirmPreference */                       typeof(ConfirmImpact),
+            /* PSNativeCommandUseErrorActionPreference */ typeof(bool),
```
Comment (iSazonov) [inline src/System.Management.Automation/engine/SpecialVariables.cs:350]: Should we follow history pattern and do not use "PS" prefix?

File: src/System.Management.Automation/resources/CommandBaseStrings.resx (hunk 9 above)
Comment (iSazonov) [reply src/System.Management.Automation/resources/CommandBaseStrings.resx:159]: I think we need to open new issue in Docs repo to explain `ApplicationFailedException` (PowerShell can not run normally an non-backgroud application) vs `NativeCommandException` (PowerShell could be able to run non-backgraund application but throw due to non zero return code).

File: src/System.Management.Automation/resources/CommandBaseStrings.resx [hunk 18]
```diff
@@ -156,6 +156,15 @@
...
+    <!-- NOTE:
+      This string was added for the native command error action preference integration feature.
+      ParserStrings already declares a ProgramFailedToExecute string,
+      however that is used for ApplicationFailedExceptions thrown when the NativeCommandProcessor fails in an unexpected way.
+      In this case, we have a more specific error for the native command scenario, so the two are not conflated.
+    -->
+    <value>Program "{0}" ended with non-zero exit code {1}.</value>
```
Comment (iSazonov) [inline src/System.Management.Automation/resources/CommandBaseStrings.resx:166]:
```suggestion
    <value>Program "{0}" ended with non-zero exit code: {1}.</value>
```

File: src/System.Management.Automation/resources/RunspaceInit.resx [hunk 19]
```diff
@@ -186,6 +186,9 @@
   <data name="NestedPromptLevelDescription" xml:space="preserve">
     <value>Dictates what type of prompt should be displayed for the current nesting level</value>
   </data>
+  <data name="PSNativeCommandUseErrorActionPreferenceDescription" xml:space="preserve">
+    <value>If true, $ErrorActionPreference applies to native executables, so that non-zero exit codes will generate cmdlet-style errors governed by error action settings</value>
```
Comment (iSazonov) [inline src/System.Management.Automation/resources/RunspaceInit.resx:190]: Probably need:
```suggestion
    <value>If true, $ErrorActionPreference applies to native executables, so that non-zero exit codes will generate cmdlet-style errors governed by error action settings.</value>
```

File: test/powershell/engine/Basic/NativeCommandErrorHandling.Tests.ps1 [hunk 20]
```diff
@@ -0,0 +1,179 @@
...
+
+    Context 'PSNativeCommandUseErrorActionPreference is $true' {
+        BeforeEach {
+            $PSNativeCommandUseErrorActionPreference = $true
+        }
+
+        It 'Non-zero exit code throws teminating error for $ErrorActionPreference = ''Stop''' {
+            $ErrorActionPreference = 'Stop'
+
+            { testexe -returncode 1 } | Should -Throw -ErrorId 'ProgramFailedToComplete'
```
Comment (iSazonov) [inline test/powershell/engine/Basic/NativeCommandErrorHandling.Tests.ps1:42]: The test shows the name is not good - the utility really _executed_ and _completed_ but return non-zero return code.
I think we need better name like NativeCommandReturnNonZeroReturnCode.

File: test/powershell/engine/Basic/NativeCommandErrorHandling.Tests.ps1 [hunk 21]
```diff
@@ -0,0 +1,179 @@
...
+            param($ErrorActionPref)
+
+            $ErrorActionPreference = $ErrorActionPref
+
+            if ($ErrorActionPref -eq 'Stop') {
+                { testexe -returncode 0 } | Should -Not -Throw
+            }
+            else {
+                testexe -returncode 0 > $null
+            }
```
Comment (iSazonov) [inline test/powershell/engine/Basic/NativeCommandErrorHandling.Tests.ps1:85]: I wonder why do we need separate branch for Stop and why do we need to write to null?

File: src/System.Management.Automation/engine/MshCommandRuntime.cs (hunk 12 above)
Comment (rjmholt) [reply src/System.Management.Automation/engine/MshCommandRuntime.cs:2822]: Actually it's directly related to it. We now call this API from multiple native calls, but that parameter was only added for a specific scenario (the stderr scenario).

The actual issue is that we'll likely be removing it soon as the experimental feature that disables it is stabilised.

File: src/System.Management.Automation/engine/NativeCommandProcessor.cs (hunk 15 above)
Comment (rjmholt) [reply src/System.Management.Automation/engine/NativeCommandProcessor.cs:906]: Didn't we discuss that for local consts camelCase is preferred?

File: src/System.Management.Automation/engine/SpecialVariables.cs (hunk 17 above)
Comment (rjmholt) [reply src/System.Management.Automation/engine/SpecialVariables.cs:350]: This was decided in the RFC though

File: src/System.Management.Automation/resources/RunspaceInit.resx (hunk 19 above)
Comment (rjmholt) [reply src/System.Management.Automation/resources/RunspaceInit.resx:190]: None of the other messages here end with `.`

File: test/powershell/engine/Basic/NativeCommandErrorHandling.Tests.ps1 (hunk 20 above)
Comment (rjmholt) [reply test/powershell/engine/Basic/NativeCommandErrorHandling.Tests.ps1:42]: Yeah that's a fair point. @theJasonHelmick, @rkeithhill we should consider renaming this error ID

File: test/powershell/engine/Basic/NativeCommandErrorHandling.Tests.ps1 (hunk 21 above)
Comment (rjmholt) [reply test/powershell/engine/Basic/NativeCommandErrorHandling.Tests.ps1:85]: Yeah it looks like we should just merge these cases. Ideally we should have some test on the output and last exit code.

File: src/System.Management.Automation/engine/SpecialVariables.cs (hunk 16 above)
Comment (rjmholt) [reply src/System.Management.Automation/engine/SpecialVariables.cs:258]: All the other variables are named this way though, so convention has established this

**Comment by rjmholt** [comment]
Opened https://github.com/rkeithhill/PowerShell/pull/22

**Comment by rjmholt** [comment]
Closing in favour of https://github.com/PowerShell/PowerShell/pull/15897 because I needed to rebase the branch and can't edit this one (@rkeithhill you can turn that on in future if you'd like)

File: src/System.Management.Automation/resources/RunspaceInit.resx (hunk 19 above)
Comment (iSazonov) [reply src/System.Management.Automation/resources/RunspaceInit.resx:190]: There are both cases. We need use right. But I don't know what case is right :-)
